Barcode finder

fbacker opened the pull request:
I've added a 'barcode finder' rect that can be enabled. When doing this it will also crop the area that is scanned. It will significantly increase the read lag. Possible to set the size, color and border.

Also I've added correct static const for barcodes on android and updated the readme.

- barcode finder
- ios: crop search area
- android: crop search area
- android: add static const of barcode types
- updated readme

sahir commented:
Hey, any news when it's going to merge ? can you please provide your branch link I did not find your branch for npm install. I want to customize react native camera something like this  

![screenshot-1505813796156](https://user-images.githubusercontent.com/452706/30587019-626d6acc-9d4f-11e7-8f71-da5599b51bb7.jpg)

fbacker commented:
https://github.com/fbacker/react-native-camera/tree/barcode-finder

until this branch is merged I'm using package.json
"react-native-camera": "github:fbacker/react-native-camera#barcode-finder"

sahir commented:
Thanks 👍

GarimaMathur07 commented:
@fbacker How can i enabled rectangle using "react-native-camera": "github:fbacker/react-native-camera#barcode-finder". It displays camera without rectangle and how can i customized that rectangle styles.

truongthanh1303 commented:
I tried your implementation, but I cannot get the expected result. All children node inside Camera component does not appear.
Did you have this issue?

Thanks.

sahir commented:
@truongthanh1303 you can use topOverlay and bottomOverlay style to appeared all children node inside camera component. for example https://github.com/lwansbrough/react-native-camera/blob/master/Example/Example.js

fbacker commented:
Try it out. Made it possible to change the default finder design. Having issues starting the package Example so developing against my own project so hope I don't miss anything or f**k up something :)

#### `barcodeFinderVisible`

Displays an rectangle over the camera to show the area of barcode scanning. If this is used the actual area that is scanned in cropped to the rectangle. This can significantly increase the performance.

Adjust size and style:
`barcodeFinderWidth`,
`barcodeFinderHeight`,
`barcodeFinderStyle`

  The default viewer style has borderColor and borderWidth.


##### Make a custom barcode finder

  1. make a copy of barcode-finder.js and place it in your project
  2. add it to your project
  3. add it as a child to the Camera
```javascript
<Camera>
    <MyCustomBarcodeFinder />
</Camera>
```
**NOTE:** The scan area is cropped and as long the first to <View> components remain intact it should show the correct size.
```javascript
<View style={[styles.container]}>
    <View style={[styles.finder, this.getSizeStyles()]}>
        { place your design here }
```

truongthanh1303 commented:
@sahir Thanks for your suggestion! It works fine :)
@fbacker I chose the first solution that @sahir suggested because it's simple. But thanks for your work! 

Regards,

sahir commented:
@truongthanh1303 @fbacker I implement barcode reader https://github.com/sahir/react-native-barcode-reader successfully but facing this issue https://github.com/lwansbrough/react-native-camera/issues/476 I have also tried the solution (bind - unbind lifecycle method) but no success. I guess you also face this issue let me know if you found any solution.

fbacker commented:
When attaching the camera display I set an empty array. Then just check if it's been used. 

~~~~
barcodeReceived(e) {
      console.log('Barcode: ' + e.data);
      console.log('Type: ' + e.type);
      if(e!=null && (e.type==Camera.constants.BarCodeType.datamatrix||e.type==Camera.constants.BarCodeType.qr)){
        let code = e.data;
        if(!this._codes.includes(code)){

          if(Platform.OS === 'ios'){
            Vibration.vibrate(500, false);
          }
          else{
            Vibration.vibrate([0, 500], false);
          }

          this._codes.push(code);
          this.props.onCode(code);
        }
      }
    }
~~~~

gablorquet commented:
It seems that the viewfinder doesn't properly restrict the barcode scanning area. 

It looks like the setBarcodeFinderVisible is never invoked in Java. I'm not too familiar with Android development so I might just be missing it (still trying to figure out how to debug java code)

GarimaMathur07 commented:
@gablorquet 

https://github.com/fbacker/react-native-camera/tree/barcode-finder

until this branch is merged, you have to use package.json
"react-native-camera": "github:fbacker/react-native-camera#barcode-finder"

Use:
`barcodeFinderVisible={true}`

```
<Camera
					ref={cam => {
						this.camera = cam;
					}}
					style={styles.preview}
					aspect={this.state.camera.aspect}
					captureTarget={this.state.camera.captureTarget}
					type={this.state.camera.type}
					flashMode={this.state.camera.flashMode}
					onFocusChanged={() => {}}
					onZoomChanged={() => {}}
					defaultTouchToFocus
					mirrorImage={false}
					barcodeFinderVisible={true}
					barcodeFinderWidth={280}
					barcodeFinderHeight={220}
					barcodeFinderBorderColor="red"
					barcodeFinderBorderWidth={2}
					onBarCodeRead={this.onBarCodeRead.bind(this)}
				/>
```

gablorquet commented:
I am, expect that in the Java class RTCCameraViewManager, the added properties are never implemented thus the barcodeScannerViewFinderEnabled variable is always false. Without them, you can still get scans from outside the ViewFinder

fbacker commented:
@GarimaMathur07 if update thru git I've changed the properties
Adjust size and style:
`barcodeFinderWidth`,
`barcodeFinderHeight`,
`barcodeFinderStyle`

also a way to make the viewfinder custom. check the updated readme.md

@gablorquet you are absolutely correct. There's so many weird steps. I'll try to fix this. Having some math calculation issues. Will try to fix this tomorrow.

fbacker commented:
I've made a new commit that should solve android crop size and also a bug in setting barcode types.

from commit:

I noticed that setting barcode types happend 1 time and was triggered before the properties was passed from react. So it was always scanning all possible formats.

Also I've changed the crop size to be based on % size. This because different resolutions on cameras didn't match up with size from react.
There's an commented code section when capture image that can be used to tweak the cropsize, so we know what we actually are looking at.

wanderSX commented:
@fbacker I'm getting error after latest commit

> `Camera` has no propType for native prop `RCTCamera.barcodeFinderPercentageSize` of native type `Array`
> If you haven't changed this prop yourself, this usually means that your versions of the native code and Javascript code are out of sync. Updating both should make this error go away

gvlamadrid commented:
@fbacker @wanderSX I think its missing define barcodeFinderPercentageSize on propType https://github.com/fbacker/react-native-camera/blob/barcode-finder/index.js#L77

fbacker commented:
@wanderSX @gvlamadrid yeah, sorry about that. Missed the index.js file in commit.

Still haven't figured out how to update npm packages outside a project. So making changes in the project folder and then copy to external git project. Tried npm link and other things. But when running multiple react-native projects it always messes up the package-manager :(

ploureiro commented:
I also got codes scans outside the scanning area in iOS, here is my code:

fbacker commented:
@ploureiro try latest.

these commits handles 1. some android devices crash sometimes for some reason and 2. hopefullt an correct crop for ios.

fbacker commented:
@gablorquet @ploureiro  please try out latest version for scan area 'cropped'

Oblongmana commented:
@fbacker Just installed your latest - the iOS scanning area seems to be working correctly, ignoring stuff outside the area. Great PR, hopefully gets merged! 👍

xilibro commented:
@fbacker thanks for your code, but i found some issues when i use this code , the scan area was not specified correctly. finally  i add some code to fix it:
`[metadataOutput setMetadataObjectTypes:self.barCodeTypes];
      
        if (self.barcodeFinderVisible) {
            double cameraViewWidth = [[UIScreen mainScreen] bounds].size.width;
            double cameraViewHeight = [[UIScreen mainScreen] bounds].size.height;
            double w = cameraViewWidth * self.barcodeFinderPercentageSizeWidth;
            double h = cameraViewHeight * self.barcodeFinderPercentageSizeHeight;
            CGRect scanLimit = CGRectMake((cameraViewWidth/2)-(w/2),(cameraViewHeight/2)-(h/2), w, h);
           + CGRect visibleMetadataOutputRect = [self.previewLayer metadataOutputRectOfInterestForRect: scanLimit];
            
           + metadataOutput.rectOfInterest = visibleMetadataOutputRect;
        }
        self.metadataOutput = metadataOutput;`

could you change your code?

xilibro commented:
hi @fbacker , thanks for your code, it is very helpful, but when i use it, i found it also can not specify scanning area correctly. i think the self.metadataOutput.rectOfInterest is the key , i have add these code and make some test, it works well now. so could you make change your code?

`
          // It seems that these setting code must below startRunning, otherwise got {{0,0},{0,0}} from func metadataOutputRectOfInterestForRect:
         [self.session startRunning];
        
        // specify scanning area by yuexl 2017-12-08 15:54:56
        if (self.barcodeFinderVisible) {
            double cameraViewWidth = [[UIScreen mainScreen] bounds].size.width;
            double cameraViewHeight = [[UIScreen mainScreen] bounds].size.height;
            double w = cameraViewWidth * self.barcodeFinderPercentageSizeWidth;
            double h = cameraViewHeight * self.barcodeFinderPercentageSizeHeight;
            CGRect scanLimit = CGRectMake((cameraViewWidth/2)-(w/2),(cameraViewHeight/2)-(h/2), w, h);
            // - [self.previewLayer metadataOutputRectOfInterestForRect: scanLimit]; 
            CGRect visibleMetadataOutputRect = [self.previewLayer metadataOutputRectOfInterestForRect: scanLimit];
            self.metadataOutput.rectOfInterest = visibleMetadataOutputRect;
            NSLog(@"debug visibleMetadataOutputRect=%@", NSStringFromCGRect(visibleMetadataOutputRect));
        }
`

fbacker commented:
@xilibro thanks for the input. Will try to look at this today or Monday and try it out. Will get back when done.

xilibro commented:
so , when this pull request can be merged @lwansbrough :)

fbacker commented:
@xilibro tried your change but that breaks my scanning area. Have you tried the latest from my branch? That should work.

I'll fix the merge conflicts.

xilibro commented:
@fbacker  strange, my device is iPhone 6 Plus.  And it is not necessary setting for specifying scanning area ? 
https://stackoverflow.com/questions/32401364/how-do-i-use-the-metadataoutputrectofinterestforrect-method-and-rectofinterest-p

xilibro commented:
Another problem is ,when i use customViewFinder , it render these children twice, could you give me some help?
![image](https://user-images.githubusercontent.com/6471292/33761339-a13a4ec2-dc43-11e7-8909-38cd9e400aea.png)
`
   <Camera
        ref={cam => {
          this.camera = cam;
        }}
        style={styles.preview}
        aspect={this.state.camera.aspect}
        captureTarget={this.state.camera.captureTarget}
        type={this.state.camera.type}
        flashMode={this.state.camera.flashMode}
        onFocusChanged={() => { }}
        onZoomChanged={() => { }}
        defaultTouchToFocus
        mirrorImage={false}
        barcodeFinderVisible={true}
        onBarCodeRead={this.onBarCodeRead.bind(this)}
      >
        <BarcodeFinder width={200} height={200} style={{borderColor:'blue', borderWidth:4}}/>
      </Camera>
`
BarcodeFinder file is just a copy of barcode-finder.js

fbacker commented:
@xilibro the viewfinder is nested, maybe that?

```
<View style={[styles.container]}>
        <View style={[styles.finder, this.getSizeStyles()]}>
```

and also the rect scanning is set by:
```
[self.previewLayer metadataOutputRectOfInterestForRect: scanLimit];
```

Running iPhone 8 plus

fbacker commented:
I've made a merge with the latest from 'lwansbrough/master'.
Note here that the scanner method is changed on android again to my solution to optimize performance when using a barcode scanner view.

xilibro commented:
hi, @fbacker , thanks for your response.
1. The property "rectOfInterest" of AVCaptureMetadataOutput is the actual setting for valid scanning area, does not it?  
`
[self.previewLayer metadataOutputRectOfInterestForRect: scanLimit];
`
about above code , apple docs says:
"Converts a rectangle in the preview layer’s coordinate system to one in the coordinate system used for metadata outputs." [https://developer.apple.com/documentation/avfoundation/avcapturevideopreviewlayer/1623495-metadataoutputrectofinterestforr](url)
it looks like the method is just do a transform job, from one coordinate to another coordinate, and finally the transformed value should give to some property, eg. "rectOfInterest" .

2.About issue of custom BarcodeFinder render two times, i will do more debug, thanks.

xilibro commented:
hi @fbacker 
i have fixed issue 2:
in react-native-camera index.js  line222
`
if(this.props.barcodeFinderVisible){
      // we need % size of viewFinder
      barcodeFinderPercentageSize = [(this.props.barcodeFinderWidth/width),(this.props.barcodeFinderHeight/(height))]
      // use default design
      // if(this.props.children !== undefined){
      //   console.log('camera render with children: ',this.props.children)
      //   childs = React.Children.map(this.props.children,
      //    (child) => React.cloneElement(child, {
      //      style: this.props.barcodeFinderStyle,
      //      width: this.props.barcodeFinderWidth,
      //      height: this.props.barcodeFinderHeight
      //    })
      //   );
      // }
      // else {
      //   // use custom viewfinder design
      //   childs = <BarcodeFinder style={this.props.barcodeFinderStyle} width={this.props.barcodeFinderWidth} height={this.props.barcodeFinderHeight} />
      // }

      if(this.props.children === undefined){
         // use custom viewfinder design
         childs = <BarcodeFinder style={this.props.barcodeFinderStyle} width={this.props.barcodeFinderWidth} height={this.props.barcodeFinderHeight} />
      }
    }

    const style = [styles.base, this.props.style];
    // this.props contains children property already.
    const nativeProps = convertNativeProps(Object.assign({},this.props,{barcodeFinderPercentageSize}));
    return (<View style={{flex:1}}>
              <RCTCamera ref={CAMERA_REF} {...nativeProps} />
              {childs}
            </View>);

`

xilibro commented:
and i also paste my issue 1 latest solution, i hope you can reconsider this issue:
`

            // specify scanning area by yuexl 2017-12-08 15:54:56
            if (self.barcodeFinderVisible) {
                // ask:how could i use [[self camera] bounds] alternative UIScreen bounds, thus we can ignore if it has navigation bar, but now it does not in main thread
                double cameraViewWidth = [[UIScreen mainScreen] bounds].size.width;
                double cameraViewHeight = [[UIScreen mainScreen] bounds].size.height; // - 64; // 64 = navigation bar + status bar height
                double w = cameraViewWidth * self.barcodeFinderPercentageSizeWidth;
                double h = cameraViewHeight * self.barcodeFinderPercentageSizeHeight;
                double x = (cameraViewWidth/2)-(w/2);
                double y = ((cameraViewHeight)/2)-(h/2);
                
                // transform to camera coordinate
                double scanX = y / cameraViewHeight;
                double scanY = x / cameraViewWidth;
                double scanWidth = h / cameraViewHeight;
                double scanHeight = w / cameraViewWidth;
                
                CGRect rectOfInterest = CGRectMake(scanX, scanY, scanWidth, scanHeight);
                
                self.metadataOutput.rectOfInterest = rectOfInterest;
            }

`

sibelius commented:
Can’t we do this in JavaScript side instead? It would make this package simpler

joshhornby commented:
@fbacker In response to: 

> Note here that the scanner method is changed on android again to my solution to optimize performance when using a barcode scanner view.

Do you mean this solution?
https://github.com/lwansbrough/react-native-camera/pull/874#issuecomment-331855227

fbacker commented:
@xilibro thanks for helping out with this. Are really swamped in work. Will try to look at your changes this week.

@joshhornby Nope here. https://github.com/fbacker/react-native-camera/blob/barcode-finder/android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraViewFinder.java
        protected Void doInBackground(Void... ignored) {

fbacker commented:
@xilibro I've fixed the issue 2. Check readme how to add custom finder. I'm not using a child but a property. Now it works.

When it comes to iOS rect for drawing. I'm still lost here. If placing barcode in center of screen is scans. But increase the area you notice that it still doesn't scan.

I've added som commented code, look for the to know what to uncomment ##############################################

**Test 1.**
When taking a photo. Crop the image to where we actually want to search

**Test 2.**
Draw UI with yellow border to show the area. Then convert it. The AVCaptureMetadataOutput rectOfInterese is using a 0-1 cordinate system. So using [_previewLayer metadataOutputRectOfInterestForRect:scanLimit] to get the correct size.

----------

Still here I can't really solve it so it works matching the visual guides. Might be some size issue? Couldn't get the photo crop to work until I recalculated the actual size. So maybe AVCaptureMetadataOutput is using another size than self.previewLayer.frame.size?

I started a stackoverflow so hopefully someone can help out. Spent 10h today on this :(
https://stackoverflow.com/questions/47817616/ios-rectofinterest-avcapturemetadataoutput-ignoring

xilibro commented:
I will do some test in weekend, hope can provide some help

xilibro commented:
I think do the layout job in react-native is better than in (iOS )native code, because the react-native-camera library do lot other jobs except scanning, most in async thread. we have to understanding the author's design logic very well. this is my humble opinion :)

Unforgiven-wanda commented:
Any progress on this?
I would love to include it in my project, the simple camera view seems a little barren

fanzkday commented:
@fbacker How to add a black backgroundColor around the rectangle?

assafb81 commented:
@fbacker great addition to the package!
I am joining @fanzkday question - how can I style the backgroundColor around the rectangle.
And another question - can I style the border - for example make the edges round and bold?

fanzkday commented:
@assafb81 Ok.

fbacker commented:
@fanzkday @assafb81 Guessing that you want the actual scan area to be transparent and a background color around the actual scanning area.

To do this. Create this outside the camera component.

assafb81 commented:
@fbacker thanks for the help.
I finally implemented it a bit differently, similar to what is done in #121.
but without the graying out the area outside the rectangle.

efstathiosntonas commented:
@fbacker if i scan a barcode and leave the camera pointing to the barcode, the scanner will continuously scan the barcode. 

Is there a way to stop the scanning when the scanner successfully scans it the 1st time?

moonbrv commented:
hello @fbacker, what if we have a lot of QR codes on 1 page, I try to catch code that is in the center of `BarcodeFinder`, but scanning catch barcode that is above this one. See pic, I try to scan QR code 101007, but instead of this camera capture 101006, or even 101005. As I know camera do scan from top to bottom, this means that area of Barcode recognition is not limited by `BarcodeFinder` size. How can I say to the camera to scan only specific area?

<details>
  <summary>Click to see screenshot</summary>

  ![](https://image.ibb.co/j0bsE6/Screenshot_2018_01_14_17_46_50_318_com_testmaster.png)
</details>

fbacker commented:
I've now run my latest in prod environment, old react-native-camera crash on android no longer exists. Yaay.

It's hard for me to keep this PR alive because the repo owners never merges to master. A lot of other stuff gets implemented and keeps breaking my PR.

So I'll close this now. Been open since September :(

zedtux commented:
@sibelius you seem to be active on this project. Could you please help @fbacker, who did a nice feature implementation, that a lot of people are looking for?

sibelius commented:
sorry about this one

anybody want to bring only the native parts of this PR?

I think the js part can be done in userland

this would solve this one https://github.com/react-native-community/react-native-camera/issues/1435

leeroybrun commented:
I've started implementing this some months ago for iOS and Android on the latest version of RNCamera based on the work of @fbacker, but never had the time to completely test it as I had to switch to another project at work. But it's used in production in one of our apps for many months now and seems to work fine.

The only part I'm not 100% sure about is the adaptation of the bounding rects coordinates when we crop the scanning area. Especially on Android and when rotating the phone, something strange seemed to be going on there.

If I find the time soon I will try to finish it and send a pull request.

In the meantime you can check it out here : https://github.com/leeroybrun/react-native-camera

This adds a new `cropScanArea={[WIDTH_PERCENTAGE, HEIGHT_PERCENTAGE]}` param to `<RNCamera />`.

For example `<RNCamera [...] cropScanArea={[0.9, 0.25]}></RNCamera>` will crop the scanning area to 90% of the camera width and 25% of its height (at the center).

The UI part is left to the developer, nothing is displayed on the screen when using this param.

Here is an example of what ~ a crop of "90% of the camera width and 25% of its height" represents.

![Sans titre-1](https://user-images.githubusercontent.com/1729528/61207557-afe92000-a6f5-11e9-8ad3-957a39c608d7.png)

Limiting the area where we need to search for a barcode results in a much faster scan for the user, especially on older phones.

andresmtz98 commented:
> I've started implementing this some months ago for iOS and Android on the latest version of RNCamera based on the work of @fbacker, but never had the time to completely test it as I had to switch to another project at work. But it's used in production in one of our apps for many months now and seems to work fine.
> 
> The only part I'm not 100% sure about is the adaptation of the bounding rects coordinates when we crop the scanning area. Especially on Android and when rotating the phone, something strange seemed to be going on there.
> 
> If I find the time soon I will try to finish it and send a pull request.
> 
> In the meantime you can check it out here : https://github.com/leeroybrun/react-native-camera
> 
> This adds a new `cropScanArea={[WIDTH_PERCENTAGE, HEIGHT_PERCENTAGE]}` param to `<RNCamera />`.
> 
> For example `<RNCamera [...] cropScanArea={[0.9, 0.25]}></RNCamera>` will crop the scanning area to 90% of the camera width and 25% of its height (at the center).
> 
> The UI part is left to the developer, nothing is displayed on the screen when using this param.
> 
> Here is an example of what ~ a crop of "90% of the camera width and 25% of its height" represents.
> 
> ![Sans titre-1](https://user-images.githubusercontent.com/1729528/61207557-afe92000-a6f5-11e9-8ad3-957a39c608d7.png)
> 
> Limiting the area where we need to search for a barcode results in a much faster scan for the user, especially on older phones.

Bro, is it able to support RN .60.x? I would like to use your fork with that feature.

sibelius commented:
Can we have something like this https://github.com/react-native-community/react-native-camera/pull/1852

Just define a rect and limit the region to it?

fabriziobertoglio1987 commented:
Android https://github.com/react-native-community/react-native-camera/pull/2766

sibelius commented:
can we close this?

fabriziobertoglio1987 commented on android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraViewFinder.java:
```diff
@@ -26,8 +29,9 @@
 import com.google.zxing.MultiFormatReader;
-import com.google.zxing.PlanarYUVLuminanceSource;
+import com.google.zxing.RGBLuminanceSource;
 import com.google.zxing.Result;
```
In my pull request https://github.com/react-native-community/react-native-camera/pull/2766 for android I expose the [PlanarYUVLuminanceSource][1] to the `reactnative` developer with a method named `rectOfInterest`.

`rectOfInterest` allows the user to `set` the instance variables later then passed to [`PlanarYUVLuminanceSource`][1] constructor. 

```java
PlanarYUVLuminanceSource(byte[] yuvData, int dataWidth, int dataHeight, int left, int top, int width, int height, boolean reverseHorizontal) 
```

read more [here](https://github.com/react-native-community/react-native-camera/pull/2766#discussion_r398858333) (the linked comment is hidden, be sure to open it).

[1]: https://zxing.github.io/zxing/apidocs/com/google/zxing/PlanarYUVLuminanceSource.html

fabriziobertoglio1987 commented on android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraModule.java:
```diff
@@ -571,6 +587,70 @@ public void onShutter() {
+                    rotationIndex = 3;
+                    break;
+            }
+            // rotate
```
I believe as discussed in https://github.com/react-native-community/react-native-camera/pull/2766 it is easier to expose the method to the `react-native` developers to create the [`PlanarYUVLuminanceSource`][1] instead of handling all the different scenarios and orientations... Many users will scan barcodes only in Landscape/FrontCamera or Portrait/FrontCamera... and maybe using `com.google.zxing` libraries is not the best choice for offering all this functionalities.. so I would be happy to discuss about this, but I will focus on the android side.
Sooner or later this functionality will break and we may not have the manpower to handle the mantainance.

[1]: https://zxing.github.io/zxing/apidocs/com/google/zxing/PlanarYUVLuminanceSource.html

```C++
            switch(deviceRotation){
                case 0: // portrait
                    rotationIndex = 6;
                    break;
                case 1: // landscape left
                    rotationIndex = 1;
                    break;
                case 2: // portrait upside down

                    break;
                case 3: // landscape right
                    rotationIndex = 3;
                    break;
            }
            // rotate
            try{
                mutableImage.rotate(rotationIndex);
            } catch (MutableImage.ImageMutationFailedException e) {
                android.util.Log.e("RCTCamera","Rotate temp image",e);
                promise.reject("failed to rotate file", e);
                return;
            }
```

You can still consider working on your own library and include this functionality, as the issue here is that the owner of the code then mantains it... as other developers may not be as productive handling issues on code they did not write, having a simple api may really help a lot.

fabriziobertoglio1987 commented on android/src/main/java/com/lwansbrough/RCTCamera/RCTCameraViewFinder.java:
```diff
@@ -300,63 +310,91 @@ public void onPreviewFrame(byte[] data, Camera camera) {
-                    throw new Exception();
-                }
+                LuminanceSource source = new RGBLuminanceSource(width, height, intArray);
+                BinaryBitmap bbitmap = new BinaryBitmap(new HybridBinarizer(source));
```
Why not continue using YUV format and [PlanarYUVLuminanceSource][2] and switching to  [`RGBLuminanceSource`][1]? I am trying to understand the reason as I am also beginner in this field. Thanks a lot

[2]: https://zxing.github.io/zxing/apidocs/com/google/zxing/PlanarYUVLuminanceSource.html

[1]: https://zxing.github.io/zxing/apidocs/com/google/zxing/RGBLuminanceSource.html

renatobenks commented:
@fabriziobertoglio1987 are you working to move this forward? I didn't get you on why did you close the other issue related to this one #2766 as you mentioned above? I mean, did you stop the work over there to get this one done? let me know to think how could I help to close this

fabriziobertoglio1987 commented:
@renatobenks currently I am busy trying to solve an issue in another repository, finishing https://github.com/react-native-community/react-native-camera/pull/2766 is in my todo list. I estimate that I will start working on it in June.

If you want to continue my work on [my branch](https://github.com/fabriziobertoglio1987/react-native-camera/tree/feature/encapsulating-logic-in-area-class) (this is the [example](https://github.com/fabriziobertoglio1987/react-native-camera/tree/encapsulating-logic-in-area-class/example) used to test the functionality). 

The remaining **TODO** explained from @cristianocca in this [comment](https://github.com/react-native-community/react-native-camera/pull/2766#issuecomment-611090714) are:
 
>What's the final goal of this change?

>Me, as a developer updating RNCamera, what should I change if I'm already using the current barcode scanner behaviour? What should I do if I want to use this improvement? 

In my effort to re-write the `rectOfInterest` functionality, I had to re-write it from scratch and introduced breaking changes. The new api is built on top of [`PlanarYUVLuminanceSource`](https://zxing.github.io/zxing/apidocs/com/google/zxing/PlanarYUVLuminanceSource.html) see this [post](https://github.com/react-native-community/react-native-camera/pull/2766#issuecomment-605499719) for detailed explanation.

The previous Javascript API was build based on `iOS` and ported to `Android`.
```
rectOfInterest={{ x: scanAreaX, y: scanAreaY, width: scanAreaWidth, height: scanAreaHeight }}
```

`x` and `y` where the coordinates of the top left corner, `width` and `height` a float number ranging from `0` to `1` representing the percentage ranging from `0%` to `100%` of the width/height of the screen.

I can build the same approach on my branch/functionality:

1) I need to build a logic on the [`rotation`](https://github.com/react-native-community/react-native-camera/blob/5abbacaac4448311775c5544ea5feb3c3b5b94f9/android/src/main/java/org/reactnative/camera/RNCameraView.java#L142) parameter
2) Rotate correctly the image based on camera rotation (currently this is calculated not with the `rotation` but on other parameters)
3) Adapt the current parameters based on [`PlanarYUVLuminanceSource`](https://zxing.github.io/zxing/apidocs/com/google/zxing/PlanarYUVLuminanceSource.html) parameters to percentages

The changes will basically reflect the fact that we pass parameters like this   
```
rectOfInterest={{ x: scanAreaX, y: scanAreaY, width: scanAreaWidth, height: scanAreaHeight }}
```
instead of 
```javascript
landscape = {
  dataWidth: 1440,
  dataHeight: 1080,
  left: 250,
  top: 300,
  width: 950,
  height: 600,
}
```
Thanks a lot
